Improvement in the UI for the Blaze Overlay

AjeshRPai opened the pull request:
Part of #17912 & #17913 

## What? 
This PR adds the following implementations to the Blaze overlay content.

1️⃣ Adds a top app bar to the Overlay and navigation buttons 
2️⃣ Fixes the background colour of the Post preview and primary button in Light :☀️ and Dark Mode 🌑
3️⃣ Adds the implementation to show the featured Image on the Preview 
4️⃣ Adds the tracking logic for Overlay events

## Why?
This Overlay is the first screen shown when a user clicks at the Blaze entry point 

 ## How? 
### Important Commits 
- [Adds: the logic for showing featured image url](https://github.com/wordpress-mobile/WordPress-Android/commit/b362d1cc87f39ba2323b7612d9b1d2ed17f63451)
- [Updates: Constraint link to handle in case featured image unavailable](https://github.com/wordpress-mobile/WordPress-Android/commit/e6ca30e4af3922a7bed7ac143c7c175c78494898)
- [Adds: the logic of changing the colors in light and dark mode](https://github.com/wordpress-mobile/WordPress-Android/commit/f939cd933e567867dc9367ad7c7144a5cd740c6c)
- [Adds: logic for fetching button text for primary button in overlay](https://github.com/wordpress-mobile/WordPress-Android/commit/5700fba1b94455cf1e47c40756c1fa8605db6326)
- [Adds: tracking events to the button and overlay display event](https://github.com/wordpress-mobile/WordPress-Android/commit/9359fb7d79d657f558b7f9707d157981ac46a28b)
- [Adds: logic of dismissing the overlay and activity](https://github.com/wordpress-mobile/WordPress-Android/commit/dc5b2359131d29189f4e9180cddea7e9d5fb0ef9)
- [Adds: tracking logic for overlay dismissal](https://github.com/wordpress-mobile/WordPress-Android/commit/37e1d693bbfaf3016ad574f4de546fcca37be94f)

 ## Screenshots 
| Light Mode  |  Dark mode | Light Mode (with Post) |  Dark Mode  |
|---|---|---|---|
| ![screenshot_site_light_mode_](https://user-images.githubusercontent.com/17463767/221666704-713961c8-d7f7-4ad6-926b-6a7f96bc3a28.png) |![screenshot_site_night_mode_](https://user-images.githubusercontent.com/17463767/221666631-a806a6b0-3bfc-4ed7-9f99-815a6a2d9c16.png)|![screenshot_post_featured_image_lightMide_](https://user-images.githubusercontent.com/17463767/221667297-f4467aaf-b72a-483b-84dd-429df82cdbb4.png) | ![screenshot_night_mode_](https://user-images.githubusercontent.com/17463767/221667349-bc0716ab-a63e-43c0-87e4-53a6161cd448.png) |


## To test:

###  📝 Pre-requisites 
- Go to the jetpack app 
- Login with a wp account with Administrator privileges 
- Go to **app settings** → **Debug settings** 
- enable feature flag → `blaze`

### Test 1 - Published Post →  Blaze overlay
1. 👆🏼 Go to app →  Posts
2. 👆🏼 Go to Published Posts →  Click on More menu
3. 👆🏼 Click on **🔥 Promote with blaze ** 
4. ✅ Verify that the Overlay is shown
5. 🔵 Verify that the event is tracked : **🔵 Tracked: blaze_overlay_displayed, Properties: {"source":"posts_list"}**
6. ✅ Verify that the post title and URL, Post featured image is shown
7. 📓  If the featured image is not available, then it wont be shown 
8. 👆🏼  Click at the **Blaze this Post** button 
9. 🔵 Verify that the event is tracked: **🔵 Tracked: blaze_overlay_button_tapped, Properties: {"source":"posts_list"}**
10. 👆🏼  Click at the close(❌) icon
11. ✅ Verify that the Overlay is dismissed 
12. 🔵 Verify that the event is tracked: **🔵 Tracked: blaze_overlay_dismissed, Properties: {"source":"posts_list"}**

### Test 2 - Dashboard card →  Blaze overlay
1. 👆🏼 Go to app -> Home
2. 👆🏼 Click on **🔥 Promote with blaze card** 
3.  🔵 Verify that the event is tracked : **🔵 Tracked: blaze_overlay_displayed, Properties: {"source":"dashboard_card"}**
4. ✅ Verify that the Overlay is shown and matches the design 
5. 👆🏼  Click at the **Blaze a Post now** button 
6.  🔵 Verify that the event is tracked : **🔵 Tracked: blaze_overlay_button_tapped, Properties: {"source":"dashboard_card"}**
7. 👆🏼  Click at the back(🔙 ) icon
8.  ✅ Verify that the Overlay is dismissed 
9. 🔵 Verify that the event is tracked: **🔵 Tracked: blaze_overlay_dismissed, Properties: {"source":"dashboard_card"}**


## Regression Notes
1. Potential unintended areas of impact
Blaze overlay is not shown as per the design spec 

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing 

3. What automated tests I added (or what prevented me from doing so)
N/A

PR submission checklist:
- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Adds: Featured image url to the UI state model

AjeshRPai committed:
Adds: the logic for showing featured image url

AjeshRPai committed:
Removes: redundant val modifier

AjeshRPai committed:
Updates: Constraint link to handle in case featured image unavailable

AjeshRPai committed:
Removes: redundant import

AjeshRPai committed:
Adds: the logic of changing the colors in light and dark mode

AjeshRPai committed:
Updates: the text styling of the bulleted and title text

AjeshRPai committed:
Moves: color values to the BlazeOverlayFragment

AjeshRPai committed:
Adds: logic for fetching button text for primary button in overlay

AjeshRPai committed:
Adds: tracking events to the button and overlay display event

AjeshRPai committed:
Adds: top bar to the blaze overlay

This commit adds two types of top bar to the blaze overlay according
to the blaze flow.

Also extends the existing MainTopAppBar to have nullable parameters,
and add Top bar parameter to the MainTopAppBar

AjeshRPai committed:
Moves: Preview functions to the bottom of the class

AjeshRPai committed:
Fixes: the top bar visibility logic

AjeshRPai committed:
Fixes: formatting and removes redundant todo comment

AjeshRPai committed:
Adds: logic of dismissing the overlay and activity

AjeshRPai committed:
Adds: tracking logic for overlay dismissal

AjeshRPai committed:
Removes: Suppress warnings

AjeshRPai commented:
Hey @ovitrif 👋🏼 

I have added you as a reviewer because I have made some changes to the existing `MainTopAppBar` with this [commit](https://github.com/wordpress-mobile/WordPress-Android/pull/18018/commits/ee41f5273c78ff09b6d27618b448917b594e287f), to support the action icons at the end and make the title an optional parameter. Also, since you have a good experience with Jetpack compose. Please don't hesitate to add feedback on the other UI logic if you have any. 

Many thanks for considering my request.

shaunandrews commented:
<img width="890" alt="CleanShot 2023-02-27 at 16 14 32@2x" src="https://user-images.githubusercontent.com/191598/221687161-dab68aaa-9f01-4783-9638-395735469f07.png">

Can we make the spacing below the list consistent? The larger space (shown on the right above) is better.

--

<img width="669" alt="CleanShot 2023-02-27 at 16 15 58@2x" src="https://user-images.githubusercontent.com/191598/221687394-5dc17d3d-39dc-4e58-95cd-67debf71acfe.png">

The spacing between the post title and URL is too large. Hard to say for sure, but reducing by half might be good.

The spacing surrounding the image should be equal. Right now I see more space on the right, shown in red. And can we crop the image to be square with rounded corners?

I don't have access to my device right now, so I can't check. Can you ensure that line-heights and text wrapping look ok with the title and URL? We can remove the protocol (https://) from the URL to help keep it short.

Maybe you could share some screen shots with longer titles, longer urls (can we truncate?), no titles, or without images.

Here's the design for reference:

<img width="634" alt="CleanShot 2023-02-27 at 16 17 32@2x" src="https://user-images.githubusercontent.com/191598/221687645-e6169d02-2f23-4a73-9521-fc6ca41bcb7b.png">

--

<img width="739" alt="CleanShot 2023-02-27 at 16 17 48@2x" src="https://user-images.githubusercontent.com/191598/221687714-6c65f719-cdae-408c-81ab-d4b4b123c854.png">

The button label is bold and gray. Is this consistent with other filled buttons on Android? In my designs the label is normal weight and white.

--

<img width="279" alt="CleanShot 2023-02-27 at 16 22 59@2x" src="https://user-images.githubusercontent.com/191598/221688681-acdbcedb-0719-4a6d-8026-d6262b1a46fa.png">

I noticed the bullets are lighter than the text. Can we make the both the same?

--

<img width="533" alt="CleanShot 2023-02-27 at 16 25 24@2x" src="https://user-images.githubusercontent.com/191598/221689116-d46a9f37-4369-483f-8979-53c4d39c7138.png">

The flame graphic seems to be missing its slightly yellow/orange background tint. Instead it looks white.

AjeshRPai committed:
Fixes: bulleted text color

AjeshRPai committed:
Fixes: the Post title and URL constraints and Style

1. Fixes the padding between the title and url
2. Adds ellipsis when the title or url gets truncated

AjeshRPai committed:
Fixes: the button text color and font weight

Updates: the ImageButton.kt component to customize according to the
inputs by adding a Button class
Updates: the documentation on the Drawable and Button

AjeshRPai committed:
Fixes: the featured image size and padding between components

AjeshRPai committed:
Fixes: the content scaling of the featured image

AjeshRPai committed:
Fixes: the spacing between the button in post/site overlay state

AjeshRPai committed:
Fixes: formatting and lint errors

AjeshRPai commented:
Hey @shaunandrews 

🙇🏼  Thanks for the review, I have done the following according to your comments above. 

- [x] Fixed the color difference in the bulleted text between bullets and text
- [x] Decreased the spacing between the Post title and URL and made both be truncated with an ellipsis if the title/url becomes more than two lines. I've attached the screenshots with the title and URL long with no image and with an image. 
- [x] I have removed the protocol (https://) from the URL to help keep it short. 
- [x] Increased the spacing between the button and the bulleted text/post-preview box 
- [x] Fixed the **Blaze this Post** button color to be white and the weight to be normal 
- [x] Fixed the spacing(to be 15dp) between the **Blaze this Post** button and the component above, when post is present or not present 

## 📹 Screenshots 
| Long title and URL with Image  | Long title & URL with no image | Without any title/URL/Image  | 
|---|---|---|
|![screenshot_long_title_url_with_image_lightMode_](https://user-images.githubusercontent.com/17463767/221806506-4193feeb-5662-40c3-8174-2df0765fe7fe.png)  | ![screenshot_long_title_url_with_no_image_lightMode_](https://user-images.githubusercontent.com/17463767/221806570-b57dfba3-5400-4da3-ba38-65655f0f7b93.png) | ![screenshot_site__](https://user-images.githubusercontent.com/17463767/221803983-e2068271-3236-4398-915c-19871d7a2d6b.png)| 

>The flame graphic seems to be missing its slightly yellow/orange background tint. Instead it looks white.

I couldn't fix this issue, I need some help from you to fix this issue. I have left a comment on the design doc cEGtlsm6xijEswaKkrQrIj-fi-477:5181. In short: While exporting the asset to SVG the details for the inner flame is lost, hence making it harder to change the color. The component that we have in Figma design has a transparent inner flame I think. It would be good if you could create a separate component, and add a color to the inner flame so that I can export it.

AjeshRPai committed:
Removes: https attribute shown on preview in blaze overlay

AjeshRPai committed:
Adds: dummy featured image url to the preview

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/compose/components/ImageButton.kt:
```diff
@@ -67,18 +72,18 @@ fun ImageButton(
-                is UiString.UiStringText -> buttonText.text
-                is UiString.UiStringRes -> stringResource(id = buttonText.stringRes)
-                is UiString.UiStringResWithParams -> stringResource(id = buttonText.stringRes)
+            val buttonTextValue: String = when (button.text) {
```
@AjeshRPai I get the feeling you could DRY this by replacing it with `uiStringText(button.text)` or `button.text.asString()` (whichever you prefer) from `org.wordpress.android.ui.compose.utils` 👍🏻

AjeshRPai replied:
@ovitrif Thanks 🙇🏼 . I missed the fact that there is a compose util, I have done the suggested change with the commit [082947b](https://github.com/wordpress-mobile/WordPress-Android/pull/18018/commits/082947b9f9c9191d90a229d7209f9b9c4272a735)

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/compose/components/ImageButton.kt:
```diff
@@ -136,5 +141,36 @@ fun ImageButton(
+ */
+data class Drawable(val resId: Int, val iconSize: Dp = 24.dp, val padding: Dp = 12.dp)
+
+/**
+ * @param text The text to be displayed.
+ * @param modifier [Modifier] to apply to this layout.
+ * @param color [Color] to apply to the text. If [Color.Unspecified], and [style] has no color set,
+ * this will be [LocalContentColor].
+ * @param fontSize The size of glyphs to use when painting the text. See [TextStyle.fontSize].
+ * @param fontStyle The typeface variant to use when drawing the letters (e.g., italic).
+ * See [TextStyle.fontStyle].
+ * @param fontWeight The typeface thickness to use when painting the text (e.g., [FontWeight.Bold]).
+ * @param fontFamily The font family to be used when rendering the text. See [TextStyle.fontFamily].
+ * @param textAlign The alignment of the text within the lines of the paragraph.
+ * See [TextStyle.textAlign].
+ * @param style Style configuration for the text such as color, font, line height etc.
+ */
```
Noice 💯 🥇

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/compose/theme/AppColor.kt:
```diff
@@ -23,15 +23,6 @@ object AppColor {
     @Stable
     val Gray40 = Color(0xFF787c82)
 
-    @Stable
-    val Gray50 = Color(0x0D000000)
-
-    @Stable
-    val Gray60 = Color(0x0DFFFFFF)
-
-    @Stable
-    val Gray70 = Color(0x1C1C1E)
-
```
Thank you 🙇🏻, Good thinking! 👌🏻

ovitrif reviewed (approved):
From a developer's point of view everything LGTM 👨🏻‍🍳 

Thank you @AjeshRPai for cleaning up some parts of the code and making others more flexible 👍🏻 

I added a small comment, aside from that all good and approved, the code is clean and it works as expected :shipit: 

P.S. Regarding the [flame icon background](https://github.com/wordpress-mobile/WordPress-Android/pull/18018#issuecomment-1447827427), you probably can add a background modifier in Compose, but I'm not sure it makes sense, what if the design changes? I guess asking for the exported asset is a better idea 👍🏻

AjeshRPai committed:
Updates: button text logic with uiStringText from Compose utils

AjeshRPai committed:
Fixes: Lint error

AjeshRPai committed:
Updates: event name as per the updated project spec

zwarm reviewed (approved):
@AjeshRPai - everything works as expected. Thank you @ovitrif for the compose review - eventually Ajesh and I will feel more comfortable with these parts.

zwarm commented:
@AjeshRPai - Based on the approvals, I am going to merge this into trunk. @shaunandrews - we will take up any additional UI changes in a separate PR. 🙇

ovitrif commented:
> Thank you @ovitrif for the compose review - eventually Ajesh and I will feel more comfortable with these parts.

Thank you both for joining the Compose ride 🙇‍♂️
